File: 2024/day10/go/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

type Point struct {
	r, c int
}

var dirs = []Point{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

func main() {
	// Read input file
	exePath, _ := os.Executable()
	exeDir := filepath.Dir(exePath)
	inputPath := filepath.Join(exeDir, "..", "input.txt")

	// Try relative path if executable path doesn't work
	data, err := os.ReadFile(inputPath)
	if err != nil {
		inputPath = "../input.txt"
		data, err = os.ReadFile(inputPath)
		if err != nil {
			panic(err)
		}
	}

	input := strings.TrimSpace(string(data))

	// Parse grid
	lines := strings.Split(input, "\n")
	rows := len(lines)
	cols := len(lines[0])
	grid := make([][]int, rows)
	for i, line := range lines {
		grid[i] = make([]int, cols)
		for j, ch := range line {
			grid[i][j] = int(ch - '0')
		}
	}

	fmt.Printf("Part 1: %d\n", part1(grid, rows, cols))
	fmt.Printf("Part 2: %d\n", part2(grid, rows, cols))
}

func findTrailheads(grid [][]int, rows, cols int) []Point {
	trailheads := []Point{}
	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			if grid[r][c] == 0 {
				trailheads = append(trailheads, Point{r, c})
			}
		}
	}
	return trailheads
}

func countReachableNines(grid [][]int, rows, cols int, startR, startC int) int {
	visited := make(map[Point]bool)
	visited[Point{startR, startC}] = true
	queue := []Point{{startR, startC}}
	nines := make(map[Point]bool)

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		currentHeight := grid[current.r][current.c]

		if currentHeight == 9 {
			nines[current] = true
			continue
		}

		// Try all four directions
		for _, dir := range dirs {
			nr := current.r + dir.r
			nc := current.c + dir.c

			if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
				next := Point{nr, nc}
				if !visited[next] && grid[nr][nc] == currentHeight+1 {
					visited[next] = true
					queue = append(queue, next)
				}
			}
		}
	}

	return len(nines)
}

func part1(grid [][]int, rows, cols int) int {
	trailheads := findTrailheads(grid, rows, cols)
	totalScore := 0
	for _, th := range trailheads {
		totalScore += countReachableNines(grid, rows, cols, th.r, th.c)
	}
	return totalScore
}

func countDistinctTrails(grid [][]int, rows, cols int, r, c int) int {
	currentHeight := grid[r][c]
	if currentHeight == 9 {
		return 1
	}

	total := 0
	for _, dir := range dirs {
		nr := r + dir.r
		nc := c + dir.c

		if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
			if grid[nr][nc] == currentHeight+1 {
				total += countDistinctTrails(grid, rows, cols, nr, nc)
			}
		}
	}

	return total
}

func part2(grid [][]int, rows, cols int) int {
	trailheads := findTrailheads(grid, rows, cols)
	totalRating := 0
	for _, th := range trailheads {
		totalRating += countDistinctTrails(grid, rows, cols, th.r, th.c)
	}
	return totalRating
}

File: 2024/day10/node/solution.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const input = readFileSync(join(__dirname, '..', 'input.txt'), 'utf-8').trim();

// Parse input into grid
const lines = input.split('\n');
const grid = lines.map(line => line.split('').map(Number));
const rows = grid.length;
const cols = grid[0].length;

// Directions: up, down, left, right
const DIRS = [[-1, 0], [1, 0], [0, -1], [0, 1]];

function findTrailheads() {
  const trailheads = [];
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (grid[r][c] === 0) {
        trailheads.push([r, c]);
      }
    }
  }
  return trailheads;
}

function countReachableNines(startR, startC) {
  const visited = new Set();
  visited.add(`${startR},${startC}`);
  const queue = [[startR, startC]];
  const nines = new Set();

  while (queue.length > 0) {
    const [r, c] = queue.shift();
    const currentHeight = grid[r][c];

    if (currentHeight === 9) {
      nines.add(`${r},${c}`);
      continue;
    }

    for (const [dr, dc] of DIRS) {
      const nr = r + dr;
      const nc = c + dc;
      if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
        const key = `${nr},${nc}`;
        if (!visited.has(key) && grid[nr][nc] === currentHeight + 1) {
          visited.add(key);
          queue.push([nr, nc]);
        }
      }
    }
  }

  return nines.size;
}

// Part 1
function part1() {
  const trailheads = findTrailheads();
  let totalScore = 0;
  for (const [r, c] of trailheads) {
    totalScore += countReachableNines(r, c);
  }
  return totalScore;
}

function countDistinctTrails(startR, startC) {
  function dfs(r, c) {
    const currentHeight = grid[r][c];
    if (currentHeight === 9) {
      return 1;
    }

    let total = 0;
    for (const [dr, dc] of DIRS) {
      const nr = r + dr;
      const nc = c + dc;
      if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
        if (grid[nr][nc] === currentHeight + 1) {
          total += dfs(nr, nc);
        }
      }
    }
    return total;
  }

  return dfs(startR, startC);
}

// Part 2
function part2() {
  const trailheads = findTrailheads();
  let totalRating = 0;
  for (const [r, c] of trailheads) {
    totalRating += countDistinctTrails(r, c);
  }
  return totalRating;
}

console.log('Part 1:', part1());
console.log('Part 2:', part2());
